### PR TITLE
plotting of python fit function

### LINKED ISSUE #20005: Evaluation and plotting of fit functions in python.
Implement function evaluation and plotting described in the design document:
https://github.com/mantidproject/documents/blob/master/Design/Fit%20functions%20in%20python.md

### PR DESCRIPTION
**Pull request by KarlPalmen**
Enable the python FunctionWrappers to plot their function.

**To test:**

Build the documentation as well as Mantid. Run plot in the various manners described in ``concepts/FitFunctiondInPython.html`` for 
a FitFunctionWrapper you have created as described in the same document.
Check that all this plot fulfils to requirements in
https://github.com/mantidproject/documents/blob/master/Design/Fit%20functions%20in%20python.md

Fixes #20005 . 

**Release Notes** 
to do

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by KarlPalmen** [comment]
It appears that these conflicts arouse from the clang patch.

File: Framework/PythonInterface/mantid/fitfunctions.py [hunk 1]
```diff
@@ -150,6 +150,72 @@ def __call__(self, x, *params):
...
+           xvals = kwargs[key]
+           haveXValues = True
+        if key == "startX":
+           xMin = kwargs[key]
+           haveStartX = True
+        if key == "endX":
+           xMax = kwargs[key]
+           haveEndX = True
+        if key == "nSteps":
+           nSteps = kwargs[key]
```
Comment (mantid-roman) [inline Framework/PythonInterface/mantid/fitfunctions.py]: Need a check that `nSteps > 0`

File: Framework/PythonInterface/mantid/fitfunctions.py [hunk 2]
```diff
@@ -150,6 +150,72 @@ def __call__(self, x, *params):
...
+              extractSpectrum = True
+        if key == "xValues":
+           xvals = kwargs[key]
+           haveXValues = True
+        if key == "startX":
+           xMin = kwargs[key]
+           haveStartX = True
+        if key == "endX":
+           xMax = kwargs[key]
+           haveEndX = True
```
Comment (mantid-roman) [inline Framework/PythonInterface/mantid/fitfunctions.py]: Need a check that `xMax > xMin`

File: docs/source/concepts/FitFunctionsInPython.rst [hunk 3]
```diff
@@ -202,6 +202,22 @@ Also one can put parameters into the function when evaluating.
...
+.. code:: python
+
+   f.plot(xValues=[0,2,2.5,3,5]) # for these x-values
+   f.plot(workpace=ws) # for the x-values of workspace ws
+   f.plot(workspace=ws, workspaceIndex=i)   # for x-values of workspace index i of ws
+   f.plot(startX=xmin, endX=xmax)  # for 20 x-values between xmin and xmax
+   f.plot(startX=xmin, endX=xmax, nSteps=10) # for 10 x-values between xmin and xmax
+   f.plot(workspace=ws, startX=xmin, endX=xmax) # for x-values of ws between xmin & xmax  
+
+Owing to the way workspaces are named in python, only one plot can be shown at a time.
```
Comment (mantid-roman) [inline docs/source/concepts/FitFunctionsInPython.rst]: We can let the user specify the name of the created workspace, default will be "function".

File: Framework/PythonInterface/mantid/fitfunctions.py [hunk 4]
```diff
@@ -150,6 +150,72 @@ def __call__(self, x, *params):
...
+      elif isWorkspace:
+          xvals = ws.readX(workspaceIndex)
+          if haveStartX and haveEndX:
+              xvals = filter(inRange, xvals)
+          if extractSpectrum or (haveStartX and haveEndX):
+              spectrumWs = CreateWorkspace( DataX=xvals, DataY=xvals)
+          else:
+              spectrumWs = ws           
+      elif haveStartX and haveEndX:
+          xvals = [ (xMax - xMin)*x/(nSteps+0.0) + xMin for x in range(nSteps+1)]
```
Comment (mantid-roman) [inline Framework/PythonInterface/mantid/fitfunctions.py]: I would prefer to use `numpy.linspace` here.

**Review by mantid-roman — CHANGES_REQUESTED** [review]

File: docs/source/concepts/FitFunctionsInPython.rst (hunk 3 above)
Comment (KarlPalmen) [reply docs/source/concepts/FitFunctionsInPython.rst]: how? NB: 'function' is a python variable name. See line 217 of fitfunctions.py .

Comment (mantid-roman) [reply docs/source/concepts/FitFunctionsInPython.rst]:
```
function = CreateWorkspace(..., OutputWorkspace=user_name)
```

Comment (KarlPalmen) [reply docs/source/concepts/FitFunctionsInPython.rst]: I'll try that and if it works, I can use function.name() as the default.

**Review by mantid-roman — APPROVED** [review]

**Pull request merged**